### PR TITLE
added test rfxencode

### PR DESCRIPTION
**Pull request by jsorg71**

### CODE REVIEW

File: tests/rfxencode.c [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+    region.x = 0;
+    region.y = 0;
+    region.cx = width;
+    region.cy = height;
+
+    num_tiles_x = (width + 63) / 64;
+    num_tiles_y = (height + 63) / 64;
+
+    num_tiles = num_tiles_x * num_tiles_y;
+    tiles = (struct rfx_tile *) malloc(num_tiles * sizeof(struct rfx_tile));
```
Comment (proski) [inline tests/rfxencode.c]: Please use calloc() and remove memset().

File: tests/rfxencode.c [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+    if (tiles == NULL)
+    {
+        free(bmp_data);
+        free(out_data);
+        return 1;
+    }
+    memset(tiles, 0, num_tiles * sizeof(struct rfx_tile));
+    for (jndex = 0; jndex < num_tiles_y; jndex++)
```
Comment (proski) [inline tests/rfxencode.c]: I would prefer `x_index` and `y_index`.

File: tests/rfxencode.c [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+        return 1;
+    }
+
+    *width = dh.width;
+    *height = dh.height;
+    *bpp = dh.bpp;
+
+    file_stride_bytes = dh.width * ((dh.bpp + 7) / 8);
+    e = (4 - file_stride_bytes) & 3;
+    src8 = (unsigned char*)malloc(file_stride_bytes * 4);
```
Comment (proski) [inline tests/rfxencode.c]: Please use space after cast here and elsewhere. Maybe use astyle?

File: tests/rfxencode.c [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+    int width;
+    int height;
+    int bpp;
+    struct rfx_rect region;
+    struct rfx_tile *tiles;
+    struct rfx_tile *tile;
+
+    out_data = (char *)malloc(1024 * 1024);
+    bmp_data = (char *)malloc(10 * 1024 * 1024);
+    memset(bmp_data, 0xff, 10 * 1024 * 1024);
```
Comment (proski) [inline tests/rfxencode.c]: I would prefer to has as few magic numbers as possible. Let's use named constants.

File: tests/Makefile.am [hunk 5]
```diff
@@ -3,9 +3,15 @@ EXTRA_DIST = readme.txt
 AM_CPPFLAGS = \
   -I$(top_srcdir)/include
 
-bin_PROGRAMS = rfxcodectest
+bin_PROGRAMS = rfxcodectest rfxencode
```
Comment (proski) [inline tests/Makefile.am]: I have doubts whether those binaries need to be installed. If not, `check_PROGRAMS` should be used.

File: tests/rfxencode.c [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+                pixel =  src8[i * 3 + 0] << 0;
+                pixel |= src8[i * 3 + 1] << 8;
+                pixel |= src8[i * 3 + 2] << 16;
+                dst32[i] = pixel;
+            }
+        }
+    }
+
+    free(src8);
+    close(fd);
```
Comment (proski) [inline tests/rfxencode.c]: Trailing space is not needed.

File: tests/rfxencode.c [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+
+    num_tiles = num_tiles_x * num_tiles_y;
+    tiles = (struct rfx_tile *) calloc(num_tiles, sizeof(struct rfx_tile));
+    if (tiles == NULL)
+    {
+        free(bmp_data);
+        free(out_data);
+        return 1;
+    }
+    for (index_y = 0; index_y < num_tiles_y; index_y++)
```
Comment (proski) [inline tests/rfxencode.c]: More trailing space.

**Review by proski — COMMENTED** [review]
How about adding a BMP file? I tried yosemite.bmp from xrdp sources. Two passes produce a shorter file than one pass. Is that expected? `-1` produces shorter output. But rfxcodectest produces a much shorter output (84908 bytes vs. 107683 bytes).

**Comment by jsorg71** [comment]
I forgot that rfxcodectest can do file in and out, I thought it only did speed test.
rfxcodectest is using a custom quantifier so it will have smaller files.  rfxencode is using the default quantizer in librfxcodec.  rlgr1 does better compression ratio but rlgr3 is faster.

**Comment by proski** [comment]
Looks good to me now. Let's merge.

**Pull request merged**